### ISSUE TITLE
`no-await-in-loop` false positive on exiting code path (break, return)

### ISSUE THREAD

**Issue by fregante** [eslint/eslint#14700]
**Tell us about your environment**

I'm testing this on eslint.org’s demo page, so I'm going to skip some N/A questions.

- **ESLint Version:**  v7.28.0
- Environment: browser
- Rules: only `no-await-in-loop`

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

[eslint.org demo link only](https://eslint.org/demo#eyJ0ZXh0IjoiYXN5bmMgZnVuY3Rpb24gaW5pdCgpIHtcbiAgZm9yIChjb25zdCBhIG9mIFswLDEsMl0pIHtcbiAgICAgIGlmIChhID09PSAxKSB7XG4gICAgICAgIGF3YWl0IGZldGNoKCcuL2RhdGEuanNvbicpO1xuXHRcdHJldHVybjtcbiAgICAgIH1cbiAgfVxufVxuXG5pbml0KCk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby1hd2FpdC1pbi1sb29wIjoyfSwiZW52Ijp7ImJyb3dzZXIiOnRydWV9fX0=)

```js
async function init() {
  for (const a of [0,1,2]) {
      if (a === 1) {
        await fetch('./data.json');
		return;
      }
  }
}

init();
```

**What did you expect to happen?**

`no-await-in-loop` is supposed to prevent sequential async operations, but here the loop is about to end when the **single*& `fetch` promise is awaited.

**What actually happened? Please copy-paste the actual, raw output from ESLint.**

```
4:9 - Unexpected `await` inside a loop. (no-await-in-loop)
```

**Comment by nzakas** [comment]
This is behaving as expected. The rule just looks for an await inside of a loop. It’s doesn’t discern how the await might affect the code path, as that could get really complex. This is an edge case so I’d suggest using a disable comment.

**Comment by fregante** [comment]
I disagree that this is an edge case. This code is equivalent to

```js
async function init() {
  wasFound = false
  for (const a of [0,1,2]) {
      if (a === 1) {
                wasFound = true 
		break;
      }
  }
  if (wasFound)
    await fetch('./data.json');
}

init();
``` 

If the rule is behaving as expected, then the rule is “dumb”, i.e. the intent of a linter is to improve code, not prevent you from putting two keywords next to each other without understanding the intent or even just the resulting behavior. There’s only one promise to ever be awaited so the rule is marking it as a problem when it’s not. 

The only thing I’d possibly agree with is that catching this case would be difficult, but other than that I think this exception should be caught.

**Comment by fregante** [comment]
I think only sub-situation where this should not be caught is if there’s a try/catch inside the for loop, because if there’s an error we could be  back to the original “sequential asynchronous loop” behavior. 

```js
async function init() {
  for (const a of [0,1,2]) {
      if (a === 1) {
        try {
          await fetch('./data.json');
          return;
        } catch {
          // oops, actually continue the loop
        }
      }
  }
}

init();
```
